## Step 1: Data Preparation

In [2]:
import pandas as pd
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

## Load the dataset

In [3]:
housing = fetch_california_housing()
data = pd.DataFrame(housing.data, columns=housing.feature_names)
data['PRICE'] = housing.target

### Split the data

In [4]:
X = data.drop('PRICE', axis=1)
y = data['PRICE']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [6]:
# Step 2: Traditional Machine Learning Models
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error

### Linear Regression

In [7]:
lr = LinearRegression()
lr.fit(X_train, y_train)
lr_pred = lr.predict(X_test)
lr_mae = mean_absolute_error(y_test, lr_pred)
lr_rmse = mean_squared_error(y_test, lr_pred, squared=False)

C:\Users\Comp2\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


### Random Forest

In [8]:
rf = RandomForestRegressor()
rf.fit(X_train, y_train)
rf_pred = rf.predict(X_test)
rf_mae = mean_absolute_error(y_test, rf_pred)
rf_rmse = mean_squared_error(y_test, rf_pred, squared=False)

C:\Users\Comp2\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


## Step 3: Neural Network Implementation

In [10]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

### Normalize the data

In [11]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

### Define the neural network

In [12]:
model = Sequential([
    Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    Dense(32, activation='relu'),
    Dense(1)
])

C:\Users\Comp2\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


### Compile and train the model

In [14]:
model.compile(optimizer='adam', loss='mse')
model.fit(X_train_scaled, y_train, epochs=50, validation_split=0.2)

Epoch 1/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 1s 646us/step - loss: 2.4682 - val_loss: 0.5226
Epoch 2/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 507us/step - loss: 0.4517 - val_loss: 0.4723
Epoch 3/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 474us/step - loss: 0.4919 - val_loss: 0.3950
Epoch 4/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 486us/step - loss: 0.3896 - val_loss: 0.3854
Epoch 5/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 512us/step - loss: 0.3607 - val_loss: 0.3706
Epoch 6/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 494us/step - loss: 0.3399 - val_loss: 0.3774
Epoch 7/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 503us/step - loss: 0.3401 - val_loss: 0.3623
Epoch 8/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 485us/step - loss: 0.3226 - val_loss: 0.3465
Epoch 9/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 534us/step - loss: 0.3131 - val_loss: 0.3403
Epoch 10/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 492us/step - loss: 0.3158 - val_loss: 0.3454
Epoch 11/50
413/413 ━━━━━━━━━━━━━━━━━━━━ 0s 505us/step - loss: 0.4343 - val_loss: 0.3408
Epoch 12/50
413/413 ━━━━━━━━━━

### Evaluate the model

In [15]:
nn_pred = model.predict(X_test_scaled)
nn_mae = mean_absolute_error(y_test, nn_pred)
nn_rmse = mean_squared_error(y_test, nn_pred, squared=False)

129/129 ━━━━━━━━━━━━━━━━━━━━ 0s 328us/step


C:\Users\Comp2\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


## Step 4: Comparison

In [16]:
print(f"Linear Regression MAE: {lr_mae}, RMSE: {lr_rmse}")
print(f"Random Forest MAE: {rf_mae}, RMSE: {rf_rmse}")
print(f"Neural Network MAE: {nn_mae}, RMSE: {nn_rmse}")

Linear Regression MAE: 0.533200130495656, RMSE: 0.7455813830127761
Random Forest MAE: 0.32750093968023264, RMSE: 0.5027936350215934
Neural Network MAE: 0.3782377092429002, RMSE: 0.5392801737672307


## Explanation of Metrics and results:
MAE (Mean Absolute Error) (smaller value = better result): This metric measures the average absolute difference between the predicted values and the actual values. It gives us an idea of how much, on average, our predictions are off from the actual values.

RMSE (Root Mean Squared Error (smaller value = better result)): This metric measures the square root of the average of the squared differences between the predicted values and the actual values. It gives us an idea of the magnitude of error in our predictions, with more emphasis on larger errors due to the squaring process.

Results Interpretation:

- Linear Regression:
MAE: 0.5332
RMSE: 0.7456
Linear Regression is a simple model that tries to fit a straight line through the data. In this case, it has an MAE of about 0.5332, meaning on average, the predictions are off by about 0.5332 units from the actual values. The RMSE is 0.7456, which indicates the spread of the errors and suggests that there are some larger errors in the predictions.

- Random Forest:
MAE: 0.3275
RMSE: 0.5028
Random Forest is an ensemble method that builds multiple decision trees and combines their results. It performs better than Linear Regression here, with a lower MAE of 0.3275. This means the average prediction error is smaller. The RMSE is also lower at 0.5028, indicating fewer and smaller large errors compared to Linear Regression.

- Neural Network:
MAE: 0.3782
RMSE: 0.5393
The Neural Network is a more complex model that can capture non-linear relationships in the data. It has an MAE of 0.3782, which is higher than Random Forest but lower than Linear Regression. The RMSE of 0.5393 suggests that while it performs better than Linear Regression, it doesn't perform as well as Random Forest in terms of minimizing large errors.

## Summary:

Random Forest performed the best overall, with the lowest MAE and RMSE, indicating it made the most accurate predictions with the smallest errors.
Neural Network performed better than Linear Regression but not as well as Random Forest.
Linear Regression had the highest errors, indicating it was the least accurate model for this dataset.
In practical terms, this means that for the California Housing Dataset, Random Forest was able to capture the relationships in the data most effectively, followed by the Neural Network, and then Linear Regression. This comparison shows the importance of trying different models and comparing their performance to select the best one for your specific problem.

## When would neural networks be better than traditional machine learning methods?

As we can see, in some cases traditional machine learning approaches like random forest or linear regression can provide better results than neural networks...however:

## Neural networks often provide better results compared to traditional machine learning models like linear regression or random forests under certain conditions. Here are some factors and scenarios where neural networks might outperform other models:

1. Complex and Non-Linear Relationships:
When the data has complex, non-linear relationships: Neural networks are highly capable of capturing complex patterns and interactions between features. If your data has such characteristics, a neural network can potentially model these relationships better than linear regression or even random forests.
2. Large Amount of Data:
When there is a large amount of data: Neural networks, particularly deep learning models, perform better with large datasets. They require a significant amount of data to learn effectively and generalize well.
3. Feature Engineering:
When feature engineering is challenging: Neural networks can automatically learn representations of the data through their multiple layers, reducing the need for manual feature engineering.
4. High-Dimensional Data:
When dealing with high-dimensional data: Neural networks can handle large numbers of features and can be particularly effective in domains like image and text data where traditional models might struggle.
5. Model Flexibility:
When flexibility is needed: Neural networks can be customized with different architectures (e.g., CNNs for images, RNNs for sequences) to fit the specific nature of the problem.
Examples and Improvements for Neural Networks:
To leverage the strengths of neural networks and potentially achieve better results, consider the following:

1. Hyperparameter Tuning:
Experiment with different architectures: Try different numbers of layers and neurons.
Adjust learning rates and batch sizes: Fine-tune the optimizer settings.
2. Regularization Techniques:
Dropout: Helps prevent overfitting by randomly setting a fraction of input units to 0 at each update during training.
L2 Regularization: Adds a penalty to the loss function to reduce overfitting.
3. Advanced Architectures:
Convolutional Neural Networks (CNNs): For image data.
Recurrent Neural Networks (RNNs) or Long Short-Term Memory networks (LSTMs): For sequential data.
4. More Data:
Collect more training data: If possible, gather more data to train the neural network, which often helps in improving performance.
5. Data Augmentation:
Data augmentation: Techniques to artificially increase the size of the training set by creating modified versions of existing data (common in image and text data).